### ISSUE TITLE
trafficjam fails when host is using nftables

### ISSUE THREAD

**Issue by ksurl** [kaysond/trafficjam#6]
log:
```
iptables: No chain/target/match by that name.
ERROR: Unexpected error while adding jump rule: iptables: No chain/target/match by that name.
```

docker version: 20.10.11
host: debian 11

compose:
```
version: "3"
services:
  trafficjam:
    image: kaysond/trafficjam
    container_name: trafficjam
    cap_add:
      - NET_ADMIN
    network_mode: host
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    environment:
      - DEBUG=true
      - NETWORK=proxy
      - "WHITELIST_FILTER=label=trafficjam.enable"
      - TZ=UTC
    restart: unless-stopped
```

**Comment by ksurl** [comment]
It seems iptables is not installed on debian lxc on proxmox. I may not be able to use this for now, until I research whether iptables is compatible on lxc since it's sharing with the host (nested containers with docker)

**Comment by kaysond** [comment]
Are you trying to run trafficjam as an LXC container on proxmox? I don't see a reason why that wouldn't work as long as it supports host-mode networking and proxmox has a compatible iptables version.

It may be that the issue is the `docker` container on which trafficjam is based uses alpine. But support for lxc/proxmox is something I definitely want, so please let me know what you find

**Comment by ksurl** [comment]
I am running trafficjam as a docker container in lxc on proxmox. This may be a rarer edge case since most people just run docker on bare metal or in a vm rather than in another container layer (I used lxc since it is lighter than a vm and faster to boot/backup/restore). In the lxc shell, no iptables is installed, but it is there on the host of course in proxmox.

**Comment by kaysond** [comment]
> I am running trafficjam as a docker container in lxc on proxmox. This may be a rarer edge case since most people just run docker on bare metal or in a vm rather than in another container layer (I used lxc since it is lighter than a vm and faster to boot/backup/restore). In the lxc shell, no iptables is installed, but it is there on the host of course in proxmox.

I see. A friend of mine has the same setup. So in that case, if you're running debian in lxc, you just need to install iptables on debian and it should work just fine.

**Comment by ksurl** [comment]
if I run a plain alpine image and install iptables, it is able to list the default rules with `iptables -L`. so the trafficjam script is having issues finding a correct rule/chain to append to?

**Comment by kaysond** [comment]
On the debian lxc, can you do `iptables -t filter -L` and post the result? Based on the error, and the fact that you do actually have iptables, it seems that there isn't a `DOCKER-USER` chain from within the trafficjam container. I'm wondering if there isn't one on the host lxc either

**Comment by kaysond** [comment]
This may be a network name spacing issue since you have an extra level of "virtualization" that I haven't tested

**Comment by kaysond** [comment]
Ok so the issue isn't LXC, but rather Debian 11. Debian uses `nftables` by default now, which is not compatible with the default in alpine - `iptables`.

You switch Debian back to "legacy" iptables by running `update-alternatives --set iptables /usr/sbin/iptables-legacy`

**Comment by c-nv-s** [comment]
as nftables is successor to iptables surely the tool should strive to support it? rather than ask users to downgrade?

**Comment by ksurl** [comment]
here's the output from `iptables -L` on the host (docker chain redacted):

```
# Warning: iptables-legacy tables present, use iptables-legacy to see them
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
DOCKER-USER  all  --  anywhere             anywhere
DOCKER-ISOLATION-STAGE-1  all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
DOCKER     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
DOCKER     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain DOCKER (2 references)
target     prot opt source               destination
...

Chain DOCKER-ISOLATION-STAGE-1 (1 references)
target     prot opt source               destination
DOCKER-ISOLATION-STAGE-2  all  --  anywhere             anywhere
DROP       all  -- !172.21.0.0/16        anywhere
 qDROP       all  --  anywhere            !x.x.x.x/16
DOCKER-ISOLATION-STAGE-2  all  --  anywhere             anywhere
RETURN     all  --  anywhere             anywhere

Chain DOCKER-ISOLATION-STAGE-2 (2 references)
target     prot opt source               destination
DROP       all  --  anywhere             anywhere
DROP       all  --  anywhere             anywhere
RETURN     all  --  anywhere             anywhere

Chain DOCKER-USER (1 references)
target     prot opt source               destination
RETURN     all  --  anywhere             anywhere
```

and from `iptables-legacy -L`:

```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain TRAFFICJAM (0 references)
target     prot opt source               destination
```

**Comment by ksurl** [comment]
> as nftables is successor to iptables surely the tool should strive to support it? rather than ask users to downgrade?

would this be as simple as checking for `nftables` and using that if detected instead of calling `iptables`? It looks like syntax is pretty similar if not identical

**Comment by ksurl** [comment]
so `/sbin/iptables-nft` is included when installing `iptables` in alpine and it shows the correct tables, not the legacy tables when called like `iptables-nft -L`. it should be straightforward to add detection for this before calling `iptables`?

**Comment by kaysond** [comment]
Yes this should be easy enough to add! I'll get it in the next nightly.

Funnily enough, I recall running into this when I first started development, but nftables wasn't so prevalent and my old Ubuntu used legacy iptables so I never ran into the problem again

**Comment by ksurl** [comment]
awesome. can't wait to finally try this out

**Comment by ksurl** [comment]
pending PR https://github.com/kaysond/trafficjam/pull/10

**Comment by ksurl** [comment]
The ubuntu-latest GitHub Actions image is based on 20.04 which still has iptables as the default. I will guess that they will switch to 22.04 by October 2022 since they switched to 20.04 in October 2020. The first ubuntu that had nftables by default is 21.10, which just released.

**Comment by kaysond** [comment]
Testing is done within a container precisely to avoid problems like this :)

**Comment by kaysond** [comment]
Fixed by #10 and #11
